File: formulus/ios/Formulus/UserAppModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTReloadCommand.h>
#import "ProfileDatabaseLifecycle.h"

@interface UserAppModule : NSObject <RCTBridgeModule>
@end

@implementation UserAppModule {
  ProfileDatabaseLifecycle *_profileDatabases;
  BOOL _restartRequested;
}

RCT_EXPORT_MODULE(UserAppModule);

+ (BOOL)requiresMainQueueSetup {
  return NO;
}

- (dispatch_queue_t)methodQueue {
  static dispatch_queue_t queue;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    queue = dispatch_queue_create("org.opendataensemble.formulus.profileLifecycle", DISPATCH_QUEUE_SERIAL);
  });
  return queue;
}

- (instancetype)init {
  if ((self = [super init])) {
    // Both WatermelonDB 0.28 JSI and ObjC adapters resolve plain names here.
    NSString *documents = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject;
    if (documents) {
      _profileDatabases = [[ProfileDatabaseLifecycle alloc] initWithDirectory:documents];
    }
  }
  return self;
}

- (BOOL)checkAvailable:(RCTPromiseRejectBlock)reject {
  if (_restartRequested || !_profileDatabases) {
    reject(@"E_PROFILE_DATABASE_STATE", @"Profile runtime is restarting or its database directory is unavailable", nil);
    return NO;
  }
  return YES;
}

- (void)rejectDatabaseError:(NSError *)error reject:(RCTPromiseRejectBlock)reject {
  NSString *code = @"E_PROFILE_DATABASE_IO";
  if (error.code == ProfileDatabaseErrorName) {
    code = @"E_PROFILE_DATABASE_NAME";
  } else if (error.code == ProfileDatabaseErrorState) {
    code = @"E_PROFILE_DATABASE_STATE";
  }
  reject(code, error.localizedDescription, error);
}

RCT_EXPORT_METHOD(prepareDatabase:(NSString *)dbName
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (![self checkAvailable:reject]) {
    return;
  }
  NSError *error = nil;
  if ([_profileDatabases prepareDatabase:dbName error:&error]) {
    resolve(nil);
  } else {
    [self rejectDatabaseError:error reject:reject];
  }
}

RCT_EXPORT_METHOD(deleteDatabase:(NSString *)dbName
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (![self checkAvailable:reject]) {
    return;
  }
  NSError *error = nil;
  NSNumber *deleted = [_profileDatabases deleteDatabase:dbName error:&error];
  if (deleted) {
    resolve(deleted);
  } else {
    [self rejectDatabaseError:error reject:reject];
  }
}

RCT_EXPORT_METHOD(isDatabaseOpen:(NSString *)dbName
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (![self checkAvailable:reject]) {
    return;
  }
  NSError *error = nil;
  NSNumber *opened = [_profileDatabases isDatabaseOpen:dbName error:&error];
  if (opened) {
    resolve(opened);
  } else {
    [self rejectDatabaseError:error reject:reject];
  }
}

RCT_EXPORT_METHOD(databaseExists:(NSString *)dbName
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (![self checkAvailable:reject]) {
    return;
  }
  NSError *error = nil;
  NSNumber *exists = [_profileDatabases databaseExists:dbName error:&error];
  if (exists) {
    resolve(exists);
  } else {
    [self rejectDatabaseError:error reject:reject];
  }
}

RCT_EXPORT_METHOD(generateProfileId:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (![self checkAvailable:reject]) {
    return;
  }
  resolve([ProfileDatabaseLifecycle generateProfileId]);
}

RCT_EXPORT_METHOD(restartRuntime:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
  if (_restartRequested) {
    reject(@"E_PROFILE_RESTART", @"Profile runtime restart is already in progress", nil);
    return;
  }
  _restartRequested = YES;
  dispatch_async(dispatch_get_main_queue(), ^{
    // Acknowledge scheduling only; JS may be disposed before it receives this.
    resolve(nil);
    // Retained for pre-preparation bootstrap retry, never for profile switching.
    // RN 0.83 supports this in release; the process-wide preparation guard survives.
    RCTTriggerReloadCommandListeners(@"Formulus bootstrap retry");
  });
}
@end

File: formulus/android/app/src/main/java/org/opendataensemble/formulus/UserAppModule.kt
package org.opendataensemble.formulus

import com.facebook.react.ReactApplication
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import java.io.File
import java.io.IOException

class UserAppModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    // WatermelonDB 0.28's Java and JSI adapters both strip /databases from this path.
    private val profileDatabases = ProfileDatabaseLifecycle(
        File(reactContext.getDatabasePath("formulus.db").path.replace("/databases", "")).parentFile!!
    )
    @Volatile private var restartRequested = false

    override fun getName(): String {
        return "UserAppModule"
    }

    private fun databaseOperation(promise: Promise, operation: () -> Any?) {
        try {
            if (restartRequested) {
                throw IllegalStateException("Profile runtime restart is already in progress")
            }
            promise.resolve(operation())
        } catch (error: IllegalArgumentException) {
            promise.reject("E_PROFILE_DATABASE_NAME", error.message, error)
        } catch (error: IllegalStateException) {
            promise.reject("E_PROFILE_DATABASE_STATE", error.message, error)
        } catch (error: IOException) {
            promise.reject("E_PROFILE_DATABASE_IO", error.message, error)
        } catch (error: SecurityException) {
            promise.reject("E_PROFILE_DATABASE_IO", error.message, error)
        }
    }

    @ReactMethod
    fun prepareDatabase(dbName: String, promise: Promise) = databaseOperation(promise) {
        profileDatabases.prepareDatabase(dbName)
        null
    }

    @ReactMethod
    fun deleteDatabase(dbName: String, promise: Promise) = databaseOperation(promise) {
        profileDatabases.deleteDatabase(dbName)
    }

    @ReactMethod
    fun isDatabaseOpen(dbName: String, promise: Promise) = databaseOperation(promise) {
        profileDatabases.isDatabaseOpen(dbName)
    }

    @ReactMethod
    fun databaseExists(dbName: String, promise: Promise) = databaseOperation(promise) {
        profileDatabases.databaseExists(dbName)
    }

    @ReactMethod
    fun generateProfileId(promise: Promise) = databaseOperation(promise) {
        ProfileDatabaseLifecycle.generateProfileId()
    }

    @ReactMethod
    fun restartRuntime(promise: Promise) {
        if (restartRequested) {
            promise.reject("E_PROFILE_RESTART", "Profile runtime restart is already in progress")
            return
        }
        restartRequested = true
        UiThreadUtil.runOnUiThread {
            try {
                val application = reactApplicationContext.applicationContext as? ReactApplication
                val host = application?.reactHost
                    ?: throw IllegalStateException("ReactHost is unavailable; close and relaunch Formulus")
                // Retained for bootstrap retry before any DB preparation, not profile switches.
                // Process-wide prepared-name guards survive this release-capable RN reload.
                host.reload("Formulus bootstrap retry")
                // Acknowledges scheduling only. The old JS runtime may die before observing this.
                promise.resolve(null)
            } catch (error: Exception) {
                restartRequested = false
                promise.reject("E_PROFILE_RESTART", "Could not restart Formulus; close and relaunch the app", error)
            }
        }
    }

    @ReactMethod
    fun getVersion(promise: Promise) {
        promise.resolve("1.0.0")
    }
}
